File: elixir/raindrops/raindrops.exs
defmodule Raindrops do

  @raindrops %{
    3 => "Pling",
    5 => "Plang",
    7 => "Plong",
  }

  @doc """
  Returns a string based on raindrop factors.

  - If the number contains 3 as a prime factor, output 'Pling'.
  - If the number contains 5 as a prime factor, output 'Plang'.
  - If the number contains 7 as a prime factor, output 'Plong'.
  - If the number does not contain 3, 5, or 7 as a prime factor,
    just pass the number's digits straight through.
  """
  @spec convert(pos_integer) :: String.t
  def convert(number) do
    case (for n <- Enum.sort(Map.keys(@raindrops)), rem(number, n) == 0, into: "", do: @raindrops[n]) do
      "" -> to_string(number)
      raindrops -> raindrops
    end
  end

end

File: go/raindrops/raindrops.go
// Package raindrops converts a number to PlingPlangPlong raindrops
package raindrops

import "strconv"

type drop struct {
	factor int
	text   string
}

var raindrops = []drop{
	{3, "Pling"},
	{5, "Plang"},
	{7, "Plong"},
}

// Convert generates a string, the contents of which depend on the number's factors.
//
// - If the number has 3 as a factor, output 'Pling'.
//
// - If the number has 5 as a factor, output 'Plang'.
//
// - If the number has 7 as a factor, output 'Plong'.
//
// - If the number does not have 3, 5, or 7 as a factor, just pass the number's digits straight through.
func Convert(number int) (rain string) {
	for _, drop := range raindrops {
		if number%drop.factor == 0 {
			rain += drop.text
		}
	}

	if len(rain) == 0 {
		rain = strconv.Itoa(number)
	}

	return rain
}
